crates/wasmtime/src/ref.rs

```diff
@@ -2,224 +2,179 @@
 
 use std::any::Any;
 use std::cell::{self, RefCell};
+use std::convert::TryFrom;
 use std::fmt;
+use std::marker::PhantomData;
 use std::rc::{Rc, Weak};
-
-trait InternalRefBase: Any {
-    fn as_any(&self) -> &dyn Any;
-    fn host_info(&self) -> Option<cell::RefMut<Box<dyn Any>>>;
-    fn set_host_info(&self, info: Option<Box<dyn Any>>);
-    fn ptr_eq(&self, other: &dyn InternalRefBase) -> bool;
-}
-
-#[derive(Clone)]
-pub struct InternalRef(Rc<dyn InternalRefBase>);
-
-impl InternalRef {
-    pub fn is_ref<T: 'static>(&self) -> bool {
-        let r = self.0.as_any();
-        Any::is::<HostRef<T>>(r)
-    }
-    pub fn get_ref<T: 'static>(&self) -> HostRef<T> {
-        let r = self.0.as_any();
-        r.downcast_ref::<HostRef<T>>()
-            .expect("reference is not T type")
-            .clone()
-    }
-}
-
-struct AnyAndHostInfo {
-    any: Box<dyn Any>,
-    host_info: Option<Box<dyn Any>>,
-}
-
-#[derive(Clone)]
-pub struct OtherRef(Rc<RefCell<AnyAndHostInfo>>);
+use wasmtime_runtime::VMExternRef;
 
 /// Represents an opaque reference to any data within WebAssembly.
 #[derive(Clone)]
-pub enum ExternRef {
-    /// A reference to no data.
-    Null,
-    /// A reference to data stored internally in `wasmtime`.
-    Ref(InternalRef),
-    /// A reference to data located outside of `wasmtime`.
-    Other(OtherRef),
+pub struct ExternRef {
+    pub(crate) inner: VMExternRef,
+    pub(crate) store: Weak<crate::runtime::StoreInner>,
 }
 
 impl ExternRef {
-    /// Creates a new instance of `ExternRef` from `Box<dyn Any>`.
-    pub fn new(data: Box<dyn Any>) -> Self {
-        let info = AnyAndHostInfo {
-            any: data,
-            host_info: None,
-        };
-        ExternRef::Other(OtherRef(Rc::new(RefCell::new(info))))
+    /// Creates a new instance of `ExternRef` wrapping the given value.
+    pub fn new<T>(store: &crate::Store, value: T) -> ExternRef
+    where
+        T: 'static + Any,
+    {
+        let inner = VMExternRef::new(value);
+        let store = store.weak();
+        ExternRef { inner, store }
     }
 
     /// Creates a `Null` reference.
-    pub fn null() -> Self {
-        ExternRef::Null
+    pub fn null() -> Option<Self> {
```

Comment:
This can probably just be deleted now, it doesn't seem to serve too much purpose any more